### PR TITLE
Speed up unit tests

### LINKED ISSUE #1815: Unit Test Runtime
This spike is intended to track the running of the unit tests.  Not sure whether it's worth tracking linux vs. windows separately here.  Unit tests are currently hitting ~20 minutes to complete locally and on CircleCI checks.  Even though we're moving to GitHub actions, that shouldn't really make the problem any better or any worse.  

The intended outcome of this spike is the following:

1. A documented methodology for profiling the unit tests.
2. Unit test profiling results posted (on a quip?) as a baseline.  (I know Steve Link showed a profiling tool that I've used before that is pretty visual and useful.)
3. Recommendations provided for the top 5 unit tests to address the performance of.
4. A quick discussion concerning the above.
5. Issues filed for the agreed results of the above.

### PR DESCRIPTION
**Pull request by freddyaboulton**
### Pull Request Description
Fixes #1815 by speeding up some unit tests and splitting up some of the jobs. 

The changes follow the analysis in this [document](https://ayx.atlassian.net/wiki/spaces/PS/pages/533233674/1815+Speed+Up+Unit+Tests). 

I think this is about a >=70% speedup in the runtime of the tests.

Splitting up the jobs is not necessary. I am including it in this PR mainly to show how it could be done. I prefer to just use a beefier machine if possible rather than adding more checks. 

-----
*After creating the pull request: in order to pass the **release_notes_updated** check you will need to update the "Future Release" section of* `docs/source/release_notes.rst` *to include this pull request by adding :pr:`123`.*

### CODE REVIEW

File: Makefile
Comment (freddyaboulton) [inline Makefile:36]: These commands are no longer used.

File: evalml/automl/automl_search.py
Comment (freddyaboulton) [inline evalml/automl/automl_search.py:286]: One of the things I found is that the sleep time increases the run time for some unit tests that run many iterations. 

This is a quick way to manipulate the amount of sleep time for some tests but it's not the only one! Let me know if you think there's a better way.

File: evalml/tests/model_understanding_tests/test_partial_dependence.py
Comment (freddyaboulton) [inline evalml/tests/model_understanding_tests/test_partial_dependence.py:405]: Since the data we use is different, the expected partial dependence is now different too

File: evalml/tests/automl_tests/test_automl.py
Comment (freddyaboulton) [inline evalml/tests/automl_tests/test_automl.py:1516]: Surprisingly, `Optimizer.tell` uses up a lot of the time for tests that run automl for many iterations.

File: evalml/tests/component_tests/test_arima_regressor.py
Comment (freddyaboulton) [inline evalml/tests/component_tests/test_arima_regressor.py:100]: I'm surprised this test took ~60 seconds since the total data size is 500 rows. 

I think we should file a separate issue to see where the bottleneck is.

File: evalml/tests/model_understanding_tests/test_partial_dependence.py
Comment (freddyaboulton) [inline evalml/tests/model_understanding_tests/test_partial_dependence.py:890]: In order to speed up these tests, I used fewer unique dates. I think this points to how our approach for handling dates does not scale super well. I think we should revisit this in the near future.

File: evalml/tests/model_understanding_tests/test_permutation_importance.py
Comment (freddyaboulton) [inline evalml/tests/model_understanding_tests/test_permutation_importance.py:480]: Running fewer objectives and columns

File: evalml/tests/component_tests/test_arima_regressor.py
Comment (chukarsten) [inline evalml/tests/component_tests/test_arima_regressor.py:317]: Yea, I'm thinking just change the fixture to be 25 rows

File: evalml/tests/model_understanding_tests/test_permutation_importance.py
Comment (chukarsten) [reply evalml/tests/model_understanding_tests/test_permutation_importance.py:480]: Curious, would it be better to replace `binary_core_objectives` with something like a `binary_test_objectives`, then we could just reduce the load on all the tests at once.  Additionally, is the fixture used after you pull it out?  Seems kinda weird that lint didn't pick it up...

**Review by chukarsten — APPROVED** [review]
I really like these changes and appreciate you taking the time to trudge through them.  I think there are a few changes that can move into fixtures, which would remove a few hardcoded values from the code base.  Also maybe start to pull more focus onto conf.py as a lever to turn up thoroughness of tests.

File: evalml/tests/component_tests/test_arima_regressor.py
Comment (angela97lin) [reply evalml/tests/component_tests/test_arima_regressor.py:100]: Also maybe super optimization / nitpick but rather than having to update the number every time we want to change how much data we use, could be useful to set a variable... just in case we later decide we don't like the number 25 or something :3

File: evalml/tests/model_understanding_tests/test_partial_dependence.py
Comment (angela97lin) [reply evalml/tests/model_understanding_tests/test_partial_dependence.py:405]: Honestly, I'm still not sure I understand the usefulness of checking against `part_dep_ans`... how are we getting these numbers anyhow 🤔

File: evalml/tests/model_understanding_tests/test_partial_dependence.py
Comment (angela97lin) [reply evalml/tests/model_understanding_tests/test_partial_dependence.py:890]: Good to file an issue? :)

**Review by angela97lin — APPROVED** [review]
This looks great, and is so, so thorough!! Mostly nit-picky comments, but I'd like to revisit our idea to use `fail-fast` as False. I think it could be better to keep that as true and move the dask tests outside the matrix, if possible, given the increased number of jobs and the number of concurrent jobs we can have. Curious about your thoughts on that, but otherwise, thank you for doing this!! The increased runtimes are super exciting. 😁

File: evalml/automl/automl_search.py
Comment (freddyaboulton) [reply evalml/automl/automl_search.py:286]: This is the [thread](https://github.com/alteryx/evalml/pull/1975#discussion_r605132883) where we discussed the sleep in the dask pr. 

We picked 0.1 just cause it seemed like a good number. Decreasing the sleep time will not change the functionality especially since these tests use the sequential engine.

File: evalml/tests/model_understanding_tests/test_partial_dependence.py
Comment (freddyaboulton) [reply evalml/tests/model_understanding_tests/test_partial_dependence.py:890]: Will do!

File: evalml/tests/model_understanding_tests/test_permutation_importance.py
Comment (freddyaboulton) [reply evalml/tests/model_understanding_tests/test_permutation_importance.py:480]: Word I'd be down for changing the fixture! I just wanted to change the slowest tests first. What are good candidates for `binary_test_objectives`?

File: Makefile
Comment (freddyaboulton) [reply Makefile:50]: I think we should measure against all evalml so that way we count the code that's defined outside of `model_understanding` that's run in the tests, e.g. pipelines, components, utils

File: evalml/tests/automl_tests/test_automl.py
Comment (freddyaboulton) [reply evalml/tests/automl_tests/test_automl.py:406]: My thought was just to modify the slowest tests. Most tests run few iterations so the sleep time doesn't contribute that much to the total time. I do think it'd be nice to leave the `0.1` value in `AutoMLSearch` but then define a fixture or something that sets it small for all tests. I'm not sure what the easiest way to do that is though and that would require modifying all evalml unit tests to add the fixture.

File: evalml/tests/component_tests/test_arima_regressor.py
Comment (freddyaboulton) [reply evalml/tests/component_tests/test_arima_regressor.py:91]: Eagle eyes. Thank you hehehe

File: evalml/tests/component_tests/test_arima_regressor.py
Comment (freddyaboulton) [reply evalml/tests/component_tests/test_arima_regressor.py:100]: I changed the fixture to only have 50 rows but the test still manually splits the data into two halves for fitting and predicting. So that's why the 25 is there. Do you want me to define another fixture to represent the second half of the data?

File: .github/workflows/linux_unit_tests_with_dask.yml
Comment (freddyaboulton) [reply .github/workflows/linux_unit_tests_with_dask.yml:1]: Correct.

File: .github/workflows/linux_unit_tests_with_latest_deps.yml
Comment (freddyaboulton) [reply .github/workflows/linux_unit_tests_with_latest_deps.yml:15]: Do you just want to set `fail-fast` to `true` for all jobs? The thing I don't like about moving tests outside the test matrix is that it leads to a lot of duplicated code across the yamls because the install steps are the same. There may be a way to reuse code across actions but I'd rather just `fail-fast` to alleviate concerns over too many jobs running.

File: evalml/tests/automl_tests/test_automl.py
Comment (freddyaboulton) [reply evalml/tests/automl_tests/test_automl.py:4579]: I don't think this was actually not mocking the pipeline fit before!

File: evalml/tests/model_understanding_tests/test_permutation_importance.py
Comment (ParthivNaresh) [reply evalml/tests/model_understanding_tests/test_permutation_importance.py:480]: I think F1, Log Loss, and AUC could be good choices

**Review by ParthivNaresh — APPROVED** [review]
Looking great! The execution time looks so much better now!

File: .github/workflows/linux_nightlies.yml
Comment (freddyaboulton) [inline .github/workflows/linux_nightlies.yml:4]: These are what the nightlies look like. Before merge I'll remove the `on: pull_request` and update the `cron` to run at 3 am. I just added the `on: pull_request` to test it out. I'll also change the `ref` of the github action to say `main`.

So if this pr is merged, this is the expected behavior:

### On pushes to prs and merges to main run: 
- linux unit tests with latest versions python=3.8, core deps [True, False]. Measure coverage and upload to codecov.
- windows unit tests with python 3.8, core deps False. Don't measure coverage.
- linux unit tests with minimum versions with python 3.7, core deps False. Don't measure coverage.

### At 3am
- linux unit tests with latest versions with python [3.7, 3.9], core deps False. Don't measure coverage.
- windows unit tests with python [3.7, 3.9], core deps False. Don't measure coverage.
- Not running minimum dependency versions as a nightly because some of the minimum package versions are not available for 3.9! I think just testing minimum versions with 3.7 is fine (and that's what we do now).


Sound good? The total number of unit test checks would be 17. FYI @chukarsten @dsherry

File: .github/workflows/linux_nightlies.yml
Comment (freddyaboulton) [inline .github/workflows/linux_nightlies.yml:14]: Setting `fail-fast` to be false so that we know what all the failures are when they occur.

File: .github/workflows/linux_unit_tests_with_minimum_deps.yml
Comment (freddyaboulton) [reply .github/workflows/linux_unit_tests_with_minimum_deps.yml]: No reason! `dir` is a leftover from a previous way I tried to parametrize the matrix. I changed them all to use `command` now.

File: evalml/tests/component_tests/test_arima_regressor.py
Comment (chukarsten) [reply evalml/tests/component_tests/test_arima_regressor.py:100]: I think that might be best @freddyaboulton

File: evalml/tests/automl_tests/test_automl.py
Comment (chukarsten) [reply evalml/tests/automl_tests/test_automl.py:410]: Whatever you think is best!  I think it's weird having the sleep time to begin with, so without any context on why we have it to begin with, my natural inclination is to just drop it for everyone in the lowest touch way possible.  You ultimately make the decision on how you want to handle this, it's not blocking!

File: evalml/tests/component_tests/test_arima_regressor.py
Comment (chukarsten) [reply evalml/tests/component_tests/test_arima_regressor.py]: lol you're just moving into the context manager....it was like that before you touched it

Comment (freddyaboulton) [reply evalml/tests/component_tests/test_arima_regressor.py]: 😂  Anyways, should be an `assert` now.

File: evalml/tests/automl_tests/test_automl.py
Comment (dsherry) [inline evalml/tests/automl_tests/test_automl.py:2607]: @freddyaboulton do you think there's a way to define a fixture for automl tests which handles stuff like `_SLEEP_TIME`, mocking pipeline methods and stuff like `Optimizer.tell`, etc?

Comment (freddyaboulton) [reply evalml/tests/automl_tests/test_automl.py:2607]: @dsherry Good point! Per our discussion at stand-up today, I'll merge this as-is but then look into a way we can package our best practices into a fixture!

File: evalml/tests/model_understanding_tests/test_permutation_importance.py
Comment (bchen1116) [inline evalml/tests/model_understanding_tests/test_permutation_importance.py:512]: Would be nice to do something similar to what @chukarsten pointed out here with `multiclass_test_objectives`

**Review by bchen1116 — APPROVED** [review]
Nice cleanup! Changes look good to me!

File: evalml/tests/automl_tests/test_automl.py
Comment (dsherry) [reply evalml/tests/automl_tests/test_automl.py:2607]: Sounds great! Sorry for any confusion, this comment is definitely not blocking this PR, and we can discuss after this PR is merged.

File: evalml/tests/model_understanding_tests/test_permutation_importance.py
Comment (freddyaboulton) [reply evalml/tests/model_understanding_tests/test_permutation_importance.py]: Done!

File: evalml/tests/model_understanding_tests/test_permutation_importance.py
Comment (freddyaboulton) [reply evalml/tests/model_understanding_tests/test_permutation_importance.py]: Done!

File: evalml/tests/component_tests/test_arima_regressor.py
Comment (freddyaboulton) [reply evalml/tests/component_tests/test_arima_regressor.py]: Done!

**Pull request merged**
